sw/device/lib/aes.c

```diff
@@ -0,0 +1,100 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/aes.h"
+
+#include "aes_regs.h"  // Generated.
+#include "sw/device/lib/common.h"
+
+#define AES0_BASE_ADDR 0x40110000
+
+void aes_init(aes_cfg_t aes_cfg) {
+  // Convert key_len from bytes to onehot
+  size_t key_len_onehot = 0b001;
+  if (aes_cfg.key_len_in_bytes == 24) {
+    key_len_onehot = 0b010;
+  } else if (aes_cfg.key_len_in_bytes == 32) {
+    key_len_onehot = 0b100;
+  }
+
+  REG32(AES_CTRL(0)) =
+      aes_cfg.mode << AES_CTRL_MODE |
+      (key_len_onehot & AES_CTRL_KEY_LEN_MASK) << AES_CTRL_KEY_LEN_OFFSET |
+      aes_cfg.manual_start_trigger << AES_CTRL_MANUAL_START_TRIGGER |
+      aes_cfg.force_data_overwrite << AES_CTRL_FORCE_DATA_OVERWRITE;
+};
+
+void aes_key_put(const void *key, size_t key_len_in_bytes) {
+  const size_t key_len_in_words = key_len_in_bytes >> 2;
+
+  for (int i = 0; i < key_len_in_words; ++i) {
+    REG32(AES_KEY0(0) + i * sizeof(uint32_t)) = ((uint32_t *)key)[i];
+  }
+  for (int i = key_len_in_words; i < 8; ++i) {
+    REG32(AES_KEY0(0) + i * sizeof(uint32_t)) = 0x0;
+  }
+}
+
+void aes_data_put_wait(const void *data) {
+  // Wait for AES unit to be ready for new input data
+  aes_data_ready();
+
+  // Provide the input data
+  aes_data_put(data);
+}
+
+void aes_data_put(const void *data) {
+  for (int i = 0; i < 4; ++i) {
+    REG32(AES_DATA_IN0(0) + i * sizeof(uint32_t)) = ((uint32_t *)data)[i];
+  }
+}
+
+void aes_data_get_wait(void *data) {
+  // Wait for AES unit to have valid output data
+  aes_data_valid();
+
+  // Get the data
+  aes_data_get(data);
+}
+
+void aes_data_get(void *data) {
+  for (int i = 0; i < 4; ++i) {
+    ((uint32_t *)data)[i] = REG32(AES_DATA_OUT0(0) + i * sizeof(uint32_t));
+  }
+}
+
+void aes_data_ready(void) {
+  while (!((REG32(AES_STATUS(0)) >> AES_STATUS_INPUT_READY) & 0x1)) {
+  }
+}
+
+void aes_data_valid(void) {
+  while (!((REG32(AES_STATUS(0)) >> AES_STATUS_OUTPUT_VALID) & 0x1)) {
```

Comment:
Consider moving the loop outside of this function, so that in the caller it becomes;

while(!aes_data_valid()) {
}

Nit: maybe for consistency with other places, instead do it like this?
while (!((REG32(AES_STATUS(0)) & (0x1u << AES_STATUS_OUTPUT_VALID))

Reply:
Thanks, the first suggestion is really good. I did not think about that. The second one I don't like to much. But I like consistency. Thus I implemented both.